conjure-java-core/src/test/resources/sealedunions/sealedunions/com/palantir/product/SingleUnion.java

```diff
@@ -0,0 +1,163 @@
+package sealedunions.com.palantir.product;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSetter;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.Safe;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nonnull;
+
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        include = JsonTypeInfo.As.EXISTING_PROPERTY,
+        property = "type",
+        visible = true,
+        defaultImpl = UnknownWrapper.class)
+@JsonSubTypes(@JsonSubTypes.Type(FooWrapper.class))
+@JsonIgnoreProperties(ignoreUnknown = true)
+public sealed interface SingleUnion {
```

Comment:
I've got some open threads about some of these questions on https://quip/3YEnAt26mmtE - to summarise:

1. `what if we drop the visitor entirely?` - as a long-term state, I definitely agree we should ditch the visitor. My thinking is that it'll probably be most efficient to keep generating it for now, to make the adoption of this flag as easy as possible:
  - I want to write the error-prone rule to rewrite visitors -> switch expressions
  - I had the tantalising thought that maybe if you have sourceCompat=17, we could actually just start generating these new sealed interfaces straight away (because sealed interfaces are GA in 17, it's only the pattern matchinbg switch expressions that require the `--enable-preview` thing)... think revapi/ABI changes might derail this plan though.
2. `*Wrapper` types - 100% agree - this is just a commit I haven't made yet.
3. `Use standard records to implement the union type` - yep I'm planning on doing this too, especially because I think destructuring is coming which should make the switch statements _even nicer_ https://openjdk.org/jeps/405.
  - `drop the static factories` I think I actually wanna keep them, because I find the IntelliJ autocomplete very natural when you type `MyUnion.` and then you get a nice dropdown of all the variants.

Reply:
I also intend to add that `.Known` sub interface because i think the `.throwOnUnknown` ergonomics will be quite neat

Reply:
Re dropping the visitor: I suppose it becomes a sequencing problem -- my preference is generally to reach for the aspirational goal, which we can set in new template projects, and follow-up with the sanest possible migration path.

Re jep405: I thought this was why you were opting into experimental features on jdk17, I don't think the sealed types and such need the experimental pattern matching support otherwise, no?

> I find the IntelliJ autocomplete very natural when you type `MyUnion.`

Wouldn't that be equivalent to `new MyUnion.`? If we're using records, I'd prefer not to provide two public constructors for the same type, as that forces developers to decide between the two.

Reply:
The main reason I have been opting-in to `--enable-preview` is strictly to get https://openjdk.org/jeps/406 (pattern matching for switch).  I only just discovered https://openjdk.org/jeps/405 (Record Patterns), but I think it'll be another usability benefit!